Orm/Xtensive.Orm/Modelling/Comparison/Comparer.cs

```diff
@@ -353,68 +360,74 @@ protected IEnumerable<NodeDifference> GetPropertyDifferences(NodeDifference diff
+        var targetSize = target?.Count ?? 0;
+        var targetKeyMap = new Dictionary<string, Node>(targetSize, StringComparer.OrdinalIgnoreCase);
+        for (var index = targetSize; index-- > 0;) {
+          var node = target[index];
+          targetKeyMap.Add(GetNodeComparisonKey(node), node);
+        }
 
-        var srcCount = source!=null ? source.Count : 0;
-        var tgtCount = target!=null ? target.Count : 0;
+        foreach (var sourceItem in sourceKeyMap) {
+          if (!targetKeyMap.ContainsKey(sourceItem.Key)) {
+            var d = Visit(sourceKeyMap[sourceItem.Key], null);
+            if (d != null) {
+              difference.ItemChanges.Add((NodeDifference) d);
+            }
+          }
+        }
 
-        if (srcCount==0 && tgtCount==0)
-          return null;
-        var someItems = srcCount!=0 ? src : tgt;
-        var someItem = someItems.Cast<Node>().First();
-
-        Func<Node, Pair<Node, string>> keyExtractor =
-          n => new Pair<Node, string>(n, GetNodeComparisonKey(n));
-
-        var sourceKeyMap = src
-          .Cast<Node>()
-          .Select(keyExtractor)
-          .ToDictionary(pair => pair.Second, pair => pair.First, StringComparer.OrdinalIgnoreCase);
-        var targetKeyMap = tgt
-          .Cast<Node>()
-          .Select(keyExtractor)
-          .ToDictionary(pair => pair.Second, pair => pair.First, StringComparer.OrdinalIgnoreCase);
-
-        var sourceKeys = src.Cast<Node>().Select(n => keyExtractor(n).Second);
-        var targetKeys = tgt.Cast<Node>().Select(n => keyExtractor(n).Second);
-        var commonKeys = sourceKeys.Intersect(targetKeys, StringComparer.OrdinalIgnoreCase);
-
-        var sequence =
-          sourceKeys.Except(commonKeys, StringComparer.OrdinalIgnoreCase)
-            .Select(k => new {Index = sourceKeyMap[k].Index, Type = 0,
-              Source = sourceKeyMap[k], Target = (Node) null})
-          .Concat(commonKeys
-            .Select(k => new {Index = targetKeyMap[k].Index, Type = 1,
-              Source = sourceKeyMap[k], Target = targetKeyMap[k]}))
-          .Concat(targetKeys.Except(commonKeys, StringComparer.OrdinalIgnoreCase)
-            .Select(k => new {Index = targetKeyMap[k].Index, Type = 2,
-              Source = (Node) null, Target = targetKeyMap[k]}))
-          .OrderBy(i => i.Type!=0).ThenBy(i => i.Index).ThenBy(i => i.Type);
-
-        foreach (var i in sequence) {
-          var d = Visit(i.Source, i.Target);
-          if (d!=null)
+        foreach (var targetItem in targetKeyMap) {
+          var (s, t) = sourceKeyMap.ContainsKey(targetItem.Key)
+            ? (sourceKeyMap[targetItem.Key], targetKeyMap[targetItem.Key])
+            : (null, targetKeyMap[targetItem.Key]);
+          var d = Visit(s, t);
+          if (d != null) {
             difference.ItemChanges.Add((NodeDifference) d);
+          }
+          
         }
+        difference.ItemChanges.Sort(CompareNodeDifference);
```

Comment:
Could you explain why we need ```Sort``` and where should I look to understand its benefits?

Reply:
It is in @botinko 's commit
https://github.com/servicetitan/dataobjects-net/pull/37/commits/778fa3f77e7aaffe5ec53e298e7644218905fc2f#diff-9b46e991888b6679cac1626def04f072288c16de1f204d421926e0f5a1090bbcR416

named Simplify.
I guess in-place Sort is more efficient than LINQ `.OrderBy()` extension. It avoid intermediate allocations

Reply:
Thank you, I missed that.